libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -199,6 +203,35 @@ size_t WaveClip::GetAppendBufferLen() const
    return mSequences[0]->GetAppendBufferLen();
 }
 
+void WaveClip::OnProjectTempoChange(
+   const std::optional<double>& oldTempo, double newTempo)
+{
+   if (!mRawAudioTempo.has_value())
+      // When we have tempo detection ready (either by header-file
+      // read-up or signal analysis) we can use something smarter than that. In
+      // the meantime, use the tempo of the project when the clip is created as
+      // source tempo.
+      mRawAudioTempo = oldTempo;
+
+   if (oldTempo.has_value())
+   {
+      const auto ratioChange = *mProjectTempo / newTempo;
+      mSequenceOffset *= ratioChange;
+      mTrimLeft *= ratioChange;
+      mTrimRight *= ratioChange;
+   }
+   mProjectTempo = newTempo;
```

Comment:
Then, more than that:  I think you are making an assumption that `mProjectTempo` is a cache of the previously seen `newTempo` value, which is expected to be the same as `oldTempo`.  The cache is needed for `GetStretchRatio`.

So you can use an assertion to document this assumption.  It won't crash release code if violated, but violations might mean something plays at wrong tempo.

So, a precondition:
`assert(!(mProjectTempo.has_value() && oldTempo.has_value()) || *mProjectTempo == *oldTempo);`

Reply:
Perfect, thanks for doing my job.

Reply:
... Although, it can be that a clip (and its track) is copied from one project to another, in which case cached project tempo and `oldTempo` may differ.